[REG-1706] Playback for programmatic playback of recordings (#73)
Regression-Games/RegressionDocs: docs/gameplay-sessions/gameplay-sessions-usage.mdx

![Open the Gameplay Session Selector](img/overlay_click_load.png)
![Select the Gameplay Session](img/overlay_choose_session.png)

## Programmatically Playing a Recording

> A recorded gameplay session can also be played back programmatically, using a 

Reviewer: ```suggestion
A recorded Gameplay Session can be played back programmatically from a 
```

[Unity Test Runner Play Mode test](https://docs.unity3d.com/Packages/com.unity.test-framework@1.4/manual/index.html).

> After creating a Play Mode test file, the `RGTestUtils` class can be used to load and play a recording, which in turn

Reviewer: ```suggestion
Create a Play Mode test file and use the `RGTestUtils` class to load and play a recording. This 
```

> will also record a new session for that test.

Reviewer: ```suggestion
will also record the replay as a new Gameplay Session.
```

:::info Note

> Recording `data.zip` files are saved by default to `~/unity_videos/{Application.productName}/run_{#}/`. Make sure

Reviewer 2: This path isn't correct

Author: Fixed here now and in the other locations where I copied this from. Since the new definition is quite long, I chose to define it in one spot and then reference it in other locations

to copy your recordings into a directory within your project to ensure that they are available for other machines
that run your automated tests.

:::

```csharp
public class RecordingTest
{

    [UnityTest]
    public IEnumerator TestRecording()
    {
        // Allow error logs to be ignored

>         LogAssert.ignoreFailingMessages = true;

Reviewer 3: Blocking: As a developer, I think having `LogAssert.ignoreFailingMessages` be a requirement for using this testing library would be a deal breaker for me. There are very few generic test oracles for games, and error logs happen to be one of them. I definitely would want my CI/CD pipeline to tell me if an unexpected error was encountered during the testing session.

From the Loom video I gathered that the motivation for this was due to an error coming out of the SDK about screenshots. If this is a non-critical error, then I would recommend to change that specific error to a warning, rather than having the test case ignoring all errors coming out of Unity entirely.

Reviewer 2: Agree this shouldn't be turned on

Author: Ok so - I spent about 2 hours really digging into this, and as explained on the call, there are a couple issues that make this pretty tough to solve:

- The exception is not actually thrown; I believe under the hood, they catch this and do `Debug.LogError`
- I can't disable error logging for just this line, because it spawns a thread, and so there is no guarantee when this will run.
- There is no config for disabling certain packages or logs from being printed. I _could_ implement my own logger, but I found that it's tough to derive or chain this logger to whatever logger the project may be using, and I don't want to override their logger in the test temporarily (i.e. the approach requires overriding / setting the Debug.unityLogger field)

Because this is only relevant in Mac, I have decided to leave this, but comment it out, and leave a comment for mac users about this error. If this is ok with everyone, then awesome! Otherwise, I will spend time implementing a custom logger

Reviewer 2: Why not just wrap it in something like this

```csharp
#if UNITY_EDITOR_OSX || UNITY_RUNTIME_OSX
#endif 
```

Reviewer 2: https://docs.unity3d.com/Manual/PlatformDependentCompilation.html

Reviewer 3: @vontell I've approved this for now since if we don't have a solution on hand I think it's OK to move forward with this for now (but I still think eventually we should find a way to reduce this requirement).

As I've written in my comment below, perhaps this approach to overriding the Debug.unityLogger could be a safe/easy way to do what is needed?  https://gist.github.com/svolokh/9956459e7961fda43cd4e262df31f745

Author: Thanks for the assistance @svolokh and @RG-nAmKcAz! I've added Zack's recommendation for now, but I also am glad you created a gist with an example - this is something I will look into! I'll make a note of it and see if I can do that at some point.
